File: 02-LINKED-LISTS/design-linked-list.py
from typing import *
from ListNode import ListNode


# https://leetcode.com/problems/design-linked-list/
class MyLinkedList:

    def __init__(self):
        self.head = ListNode()
        self.tail = None
        self.size = 0

    def get(self, index: int) -> int:
        if index < 0 or index >= self.size:
            return -1

        current = self.head.next
        for i in range(index):
            current = current.next

        return current.val

    def addAtHead(self, val: int) -> None:
        newNode = ListNode(val, self.head.next, None)
        if self.head.next:
            self.head.next.prev = newNode

        if self.tail is None:
            self.tail = newNode

        self.head.next = newNode
        self.size += 1

    def addAtTail(self, val: int) -> None:
        newNode = ListNode(val, None, self.tail)

        if self.tail:
            self.tail.next = newNode
        else:
            newNode.prev = self.head
            self.head.next = newNode

        self.tail = newNode
        self.size += 1

    def addAtIndex(self, index: int, val: int) -> None:
        if index < 0 or index > self.size:
            return

        if index == 0:
            self.addAtHead(val)
        elif index == self.size:
            self.addAtTail(val)
        else:
            current = self.head.next
            for i in range(index - 1):
                current = current.next

            newNode = ListNode(val, current.next, current)
            current.next.prev = newNode
            current.next = newNode
            self.size += 1

    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.size:
            return

        current = self.head.next
        for i in range(index):
            current = current.next

        if current.prev:
            current.prev.next = current.next
        else:
            self.head.next = current.next

        if current.next:
            current.next.prev = current.prev
        else:
            self.tail = current.prev

        self.size -= 1

File: 02-LINKED-LISTS/PHP/design-linked-list.php
<?php
use ListNode as MyListNode;

# https://leetcode.com/problems/design-linked-list/
class MyLinkedList {
    public ?MyListNode $head;
    public ?MyListNode $tail;
    public int $size;

    public function __construct() {
        $this->head = new MyListNode();
        $this->tail = null;
        $this->size = 0;
    }

    public function get(int $index): int
    {
        if ($index < 0 || $index >= $this->size) {
            return -1;
        }

        $current = $this->head->next;
        for ($i = 0; $i < $index; $i++) {
            $current = $current->next;
        }

        return $current->val;
    }

    public function addAtHead(int $val): void
    {
        $newNode = new MyListNode($val, $this->head->next, null);
        if ($this->head->next) {
            $this->head->next->prev = $newNode;
        }

        if ($this->tail === null) {
            $this->tail = $newNode;
        }

        $this->head->next = $newNode;
        $this->size++;
    }

    public function addAtTail(int $val): void
    {
        $newNode = new MyListNode($val, null, $this->tail);

        if ($this->tail) {
            $this->tail->next = $newNode;
        } else {
            $newNode->prev = $this->head;
            $this->head->next = $newNode;
        }

        $this->tail = $newNode;
        $this->size++;
    }

    public function addAtIndex(int $index, int $val): void
    {
        if ($index < 0 || $index > $this->size) {
            return;
        }

        if ($index === 0) {
            $this->addAtHead($val);
        } elseif ($index === $this->size) {
            $this->addAtTail($val);
        } else {
            $current = $this->head->next;
            for ($i = 0; $i < $index - 1; $i++) {
                $current = $current->next;
            }

            $newNode = new MyListNode($val, $current->next, $current);
            $current->next->prev = $newNode;
            $current->next = $newNode;
            $this->size++;
        }
    }

    public function deleteAtIndex(int $index): void
    {
        if ($index < 0 || $index >= $this->size) {
            return;
        }

        $current = $this->head->next;
        for ($i = 0; $i < $index; $i++) {
            $current = $current->next;
        }

        if ($current->prev) {
            $current->prev->next = $current->next;
        } else {
            $this->head->next = $current->next;
        }

        if ($current->next) {
            $current->next->prev = $current->prev;
        } else {
            $this->tail = $current->prev;
        }

        $this->size--;
    }
}
